tools/loki.go

```diff
@@ -0,0 +1,484 @@
+	StartRFC3339  string `json:"startRfc3339,omitempty" jsonschema:"description=Optionally, the start time of the query in RFC3339 format"`
+	EndRFC3339    string `json:"endRfc3339,omitempty" jsonschema:"description=Optionally, the end time of the query in RFC3339 format"`
+	Limit         int    `json:"limit,omitempty" jsonschema:"description=Optionally, the maximum number of log lines to return (default: 10, max: 100)"`
+	Direction     string `json:"direction,omitempty" jsonschema:"description=Optionally, the direction of the query: 'forward' (oldest first) or 'backward' (newest first, default)"`
+}
+
+// LogEntry represents a single log entry with metadata
+type LogEntry struct {
+	Timestamp string            `json:"timestamp"`
+	Line      string            `json:"line"`
+	Labels    map[string]string `json:"labels"`
+}
+
+// enforceLogLimit ensures a log limit value is within acceptable bounds
+func enforceLogLimit(requestedLimit int) int {
+	if requestedLimit <= 0 {
+		return DefaultLokiLogLimit
+	}
+	if requestedLimit > MaxLokiLogLimit {
+		return MaxLokiLogLimit
+	}
+	return requestedLimit
+}
+
+// queryLokiLogs queries logs from a Loki datasource using LogQL
+func queryLokiLogs(ctx context.Context, args QueryLokiLogsParams) ([]LogEntry, error) {
+	client, err := newLokiClient(ctx, args.DatasourceUID)
+	if err != nil {
+		return nil, fmt.Errorf("creating Loki client: %w", err)
+	}
+
+	// Get default time range if not provided
+	startTime, endTime := getDefaultTimeRange(args.StartRFC3339, args.EndRFC3339)
+
+	// Apply limit constraints
+	limit := enforceLogLimit(args.Limit)
+
+	// Set default direction if not provided
+	direction := args.Direction
+	if direction == "" {
+		direction = "backward" // Most recent logs first
+	}
+
+	streams, err := client.fetchLogs(ctx, args.LogQL, startTime, endTime, limit, direction)
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle empty results
+	if len(streams) == 0 {
+		return []LogEntry{}, nil
+	}
+
+	// Convert the streams to a flat list of log entries
+	var entries []LogEntry
+	for _, stream := range streams {
+		for _, value := range stream.Values {
+			if len(value) >= 2 {
+				entry := LogEntry{
+					Timestamp: value[0],
+					Line:      value[1],
+					Labels:    stream.Stream,
+				}
+				entries = append(entries, entry)
+			}
+		}
+	}
+
+	// If we processed all streams but still have no entries, return an empty slice
+	if len(entries) == 0 {
+		return []LogEntry{}, nil
+	}
+
+	return entries, nil
+}
+
+// QueryLokiLogs is a tool for querying logs from Loki
+var QueryLokiLogs = mcpgrafana.MustTool(
+	"query_loki_logs",
+	"Query and retrieve log entries from a Loki datasource using LogQL. Returns log lines with their timestamps and associated label metadata. Supports full LogQL syntax including filters and expressions.",
```

Comment:
It might be worth adding something like "Before using this tool, use `query_loki_stats` to ensure the log stream is not too large, `list_loki_label_names` to ensure the labels exist, and `list_loki_label_values` to ensure the label values exist.